File: libraries/python-types/RiskObject.py
"""
RiskObject - Core risk data structure for CyberRX platform.

This is the canonical representation of risk that flows through the entire system.
All connectors normalize their data to this schema.

CRITICAL: This is the core data structure. Get this right, everything else becomes easier.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import List, Literal, Optional
from enum import Enum
from uuid import UUID


class RiskCategory(str, Enum):
    """Risk category classification."""
    THREAT = "threat"
    VULNERABILITY = "vulnerability"
    COMPLIANCE = "compliance"
    VENDOR = "vendor"
    OPERATIONAL = "operational"


class RiskStatus(str, Enum):
    """Current status of the risk."""
    ACTIVE = "active"
    REMEDIATED = "remediated"
    ACCEPTED = "accepted"
    ESCALATED = "escalated"


@dataclass
class MethodologyTrail:
    """Audit trail for how a risk object was created and enriched.

    CRITICAL: This enables CFO board-meeting defensibility by tracking
    every step of the normalization and enrichment process.
    """
    normalization_steps: List[str]
    enrichment_timestamps: List[str]
    data_sources: List[str]
    calculation_methods: List[str]
    assumptions: List[str]
    confidence_scores: List[float]

    def to_dict(self) -> dict:
        return {
            "normalization_steps": self.normalization_steps,
            "enrichment_timestamps": self.enrichment_timestamps,
            "data_sources": self.data_sources,
            "calculation_methods": self.calculation_methods,
            "assumptions": self.assumptions,
            "confidence_scores": self.confidence_scores
        }


@dataclass
class Regulation:
    """Regulatory trigger associated with a risk."""
    regulation_id: str
    name: str
    obligation: str
    deadline: str  # ISO 8601
    status: Literal["compliant", "at_risk", "non_compliant"]
    notification_required: bool
    notification_timeline: str  # e.g., "60 days"
    cms_form_required: Optional[str] = None  # e.g., "CMS-10743"

    def to_dict(self) -> dict:
        return {
            "regulation_id": self.regulation_id,
            "name": self.name,
            "obligation": self.obligation,
            "deadline": self.deadline,
            "status": self.status,
            "notification_required": self.notification_required,
            "notification_timeline": self.notification_timeline,
            "cms_form_required": self.cms_form_required
        }


@dataclass
class Threshold:
    """Threshold breach associated with a risk."""
    threshold_id: str
    threshold_name: str
    threshold_type: Literal["risk_score", "financial_exposure", "trend", "custom"]
    threshold_value: float
    actual_value: float
    severity: Literal["low", "medium", "high", "critical"]
    triggered_at: str  # ISO 8601
    acknowledged: bool = False
    acknowledged_by: Optional[str] = None
    acknowledged_at: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "threshold_id": self.threshold_id,
            "threshold_name": self.threshold_name,
            "threshold_type": self.threshold_type,
            "threshold_value": self.threshold_value,
            "actual_value": self.actual_value,
            "severity": self.severity,
            "triggered_at": self.triggered_at,
            "acknowledged": self.acknowledged,
            "acknowledged_by": self.acknowledged_by,
            "acknowledged_at": self.acknowledged_at
        }


@dataclass
class RiskObject:
    """
    Core risk data structure that flows through the entire CyberRX system.

    This is the canonical representation of risk. All connectors normalize
    their data to this schema before publishing to the event bus.

    CRITICAL FIELDS:
    - financial_exposure: Must be CFO board-meeting defensible
    - methodology_trail: Complete audit trail for all calculations
    - business_process_map: Links to business process graph for impact analysis
    - blast_radius: Downstream systems reachable by this risk
    """

    # Identity
    id: UUID
    source: str  # Connector identifier (e.g., "splunk", "crowdstrike")
    source_event_id: str  # Original event ID from source
    category: RiskCategory

    # What's affected
    affected_assets: List[str]  # System names, hostnames, IPs
    business_process_map: List[str]  # Business process IDs

    # Risk assessment
    likelihood_score: float  # 0.0 - 1.0
    blast_radius: List[str]  # Downstream systems reachable
    financial_exposure: 'FinancialImpact'
    regulatory_triggers: List[Regulation]
    threshold_breaches: List[Threshold]

    # Resolution
    remediation_owner: str  # Team or person responsible
    status: RiskStatus

    # Metadata
    created_at: str  # ISO 8601
    updated_at: str  # ISO 8601
    first_detected_at: str  # ISO 8601
    confidence: float  # 0.0 - 1.0

    # Audit trail
    methodology_trail: MethodologyTrail
    normalization_notes: str

    def to_dict(self) -> dict:
        """Convert RiskObject to dictionary for JSON serialization."""
        return {
            "id": str(self.id),
            "source": self.source,
            "source_event_id": self.source_event_id,
            "category": self.category.value,
            "affected_assets": self.affected_assets,
            "business_process_map": self.business_process_map,
            "likelihood_score": self.likelihood_score,
            "blast_radius": self.blast_radius,
            "financial_exposure": self.financial_exposure.to_dict(),
            "regulatory_triggers": [r.to_dict() for r in self.regulatory_triggers],
            "threshold_breaches": [t.to_dict() for t in self.threshold_breaches],
            "remediation_owner": self.remediation_owner,
            "status": self.status.value,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "first_detected_at": self.first_detected_at,
            "confidence": self.confidence,
            "methodology_trail": self.methodology_trail.to_dict(),
            "normalization_notes": self.normalization_notes
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'RiskObject':
        """Create RiskObject from dictionary."""
        from .FinancialImpact import FinancialImpact

        financial_exposure = FinancialImpact.from_dict(data["financial_exposure"])
        regulatory_triggers = [Regulation(**r) for r in data["regulatory_triggers"]]
        threshold_breaches = [Threshold(**t) for t in data["threshold_breaches"]]
        methodology_trail = MethodologyTrail(**data["methodology_trail"])

        return cls(
            id=UUID(data["id"]),
            source=data["source"],
            source_event_id=data["source_event_id"],
            category=RiskCategory(data["category"]),
            affected_assets=data["affected_assets"],
            business_process_map=data["business_process_map"],
            likelihood_score=data["likelihood_score"],
            blast_radius=data["blast_radius"],
            financial_exposure=financial_exposure,
            regulatory_triggers=regulatory_triggers,
            threshold_breaches=threshold_breaches,
            remediation_owner=data["remediation_owner"],
            status=RiskStatus(data["status"]),
            created_at=data["created_at"],
            updated_at=data["updated_at"],
            first_detected_at=data["first_detected_at"],
            confidence=data["confidence"],
            methodology_trail=methodology_trail,
            normalization_notes=data["normalization_notes"]
        )

    def validate(self) -> bool:
        """Validate RiskObject constraints."""
        # Check score ranges
        if not (0.0 <= self.likelihood_score <= 1.0):
            raise ValueError(f"Invalid likelihood_score: {self.likelihood_score}")
        if not (0.0 <= self.confidence <= 1.0):
            raise ValueError(f"Invalid confidence: {self.confidence}")

        # Check required fields
        if not self.affected_assets:
            raise ValueError("affected_assets cannot be empty")
        if not self.business_process_map:
            raise ValueError("business_process_map cannot be empty")
        if not self.remediation_owner:
            raise ValueError("remediation_owner cannot be empty")

        return True

File: libraries/types/RiskObject.ts
/**
 * RiskObject - Core risk data structure for CyberRX platform.
 *
 * This is the canonical representation of risk that flows through the entire system.
 * All connectors normalize their data to this schema before publishing to the event bus.
 *
 * CRITICAL: This is the core data structure. Get this right, everything else becomes easier.
 *
 * @packageDocumentation
 */

/**
 * Risk category classification.
 */
export enum RiskCategory {
  THREAT = "threat",
  VULNERABILITY = "vulnerability",
  COMPLIANCE = "compliance",
  VENDOR = "vendor",
  OPERATIONAL = "operational"
}

/**
 * Current status of the risk.
 */
export enum RiskStatus {
  ACTIVE = "active",
  REMEDIATED = "remediated",
  ACCEPTED = "accepted",
  ESCALATED = "escalated"
}

/**
 * Audit trail for how a risk object was created and enriched.
 *
 * CRITICAL: This enables CFO board-meeting defensibility by tracking
 * every step of the normalization and enrichment process.
 */
export interface MethodologyTrail {
  /** Sequential steps taken to normalize this risk */
  normalization_steps: string[];
  /** Timestamps for each enrichment operation */
  enrichment_timestamps: string[];
  /** Data sources that contributed to this risk object */
  data_sources: string[];
  /** Calculation methods used (e.g., "blast_radius_algorithm_v1") */
  calculation_methods: string[];
  /** Key assumptions made during normalization */
  assumptions: string[];
  /** Confidence scores for each major step */
  confidence_scores: number[];
}

/**
 * Regulatory trigger associated with a risk.
 */
export interface Regulation {
  /** Unique regulation identifier (e.g., "CMS-10743") */
  regulation_id: string;
  /** Regulation name (e.g., "HIPAA 45 CFR §164.312") */
  name: string;
  /** Specific obligation triggered by this risk */
  obligation: string;
  /** Deadline for compliance (ISO 8601) */
  deadline: string;
  /** Current compliance status */
  status: "compliant" | "at_risk" | "non_compliant";
  /** Whether notification to regulators is required */
  notification_required: boolean;
  /** Timeline for notification (e.g., "60 days") */
  notification_timeline: string;
  /** CMS form required if applicable (e.g., "CMS-10743") */
  cms_form_required?: string;
}

/**
 * Threshold breach associated with a risk.
 */
export interface Threshold {
  /** Unique threshold identifier */
  threshold_id: string;
  /** Human-readable threshold name */
  threshold_name: string;
  /** Type of threshold breached */
  threshold_type: "risk_score" | "financial_exposure" | "trend" | "custom";
  /** Threshold value that was breached */
  threshold_value: number;
  /** Actual value that triggered the breach */
  actual_value: number;
  /** Severity of the breach */
  severity: "low" | "medium" | "high" | "critical";
  /** When the threshold was breached (ISO 8601) */
  triggered_at: string;
  /** Whether the breach has been acknowledged */
  acknowledged: boolean;
  /** Who acknowledged the breach */
  acknowledged_by?: string;
  /** When the breach was acknowledged (ISO 8601) */
  acknowledged_at?: string;
}

/**
 * Core risk data structure that flows through the entire CyberRX system.
 *
 * This is the canonical representation of risk. All connectors normalize
 * their data to this schema before publishing to the event bus.
 *
 * CRITICAL FIELDS:
 * - financial_exposure: Must be CFO board-meeting defensible
 * - methodology_trail: Complete audit trail for all calculations
 * - business_process_map: Links to business process graph for impact analysis
 * - blast_radius: Downstream systems reachable by this risk
 *
 * @example
 * ```typescript
 * const riskObject: RiskObject = {
 *   id: "uuid",
 *   source: "splunk",
 *   source_event_id: "splunk-event-123",
 *   category: RiskCategory.THREAT,
 *   affected_assets: ["server-1", "server-2"],
 *   business_process_map: ["claims-adjudication"],
 *   likelihood_score: 0.8,
 *   blast_radius: ["database-1", "edi-gateway"],
 *   financial_exposure: { ... },
 *   regulatory_triggers: [ ... ],
 *   threshold_breaches: [ ... ],
 *   remediation_owner: "security-team",
 *   status: RiskStatus.ACTIVE,
 *   created_at: "2025-06-05T12:00:00Z",
 *   updated_at: "2025-06-05T12:00:00Z",
 *   first_detected_at: "2025-06-05T12:00:00Z",
 *   confidence: 0.9,
 *   methodology_trail: { ... },
 *   normalization_notes: "Normalized from Splunk alert format"
 * };
 * ```
 */
export interface RiskObject {
  // Identity
  /** Unique identifier (UUID) */
  id: string;
  /** Connector that produced this risk (e.g., "splunk", "crowdstrike") */
  source: string;
  /** Original event ID from source system */
  source_event_id: string;
  /** Risk category classification */
  category: RiskCategory;

  // What's affected
  /** Systems, assets, hostnames, IPs affected by this risk */
  affected_assets: string[];
  /** Business process IDs impacted (links to BusinessProcessGraph) */
  business_process_map: string[];

  // Risk assessment
  /** Probability of exploitation (0.0 - 1.0) */
  likelihood_score: number;
  /** Downstream systems reachable from affected assets */
  blast_radius: string[];
  /** Financial impact calculation (CFO board-meeting defensible) */
  financial_exposure: FinancialImpact;
  /** Regulatory obligations triggered by this risk */
  regulatory_triggers: Regulation[];
  /** Thresholds breached by this risk */
  threshold_breaches: Threshold[];

  // Resolution
  /** Team or person responsible for remediation */
  remediation_owner: string;
  /** Current status of the risk */
  status: RiskStatus;

  // Metadata
  /** When this risk object was created (ISO 8601) */
  created_at: string;
  /** When this risk object was last updated (ISO 8601) */
  updated_at: string;
  /** When the risk was first detected (ISO 8601) */
  first_detected_at: string;
  /** Confidence in the assessment (0.0 - 1.0) */
  confidence: number;

  // Audit trail
  /** Complete methodology trail for CFO defensibility */
  methodology_trail: MethodologyTrail;
  /** Notes on normalization from source format */
  normalization_notes: string;
}

/**
 * Forward declaration for FinancialImpact.
 * This will be imported from the FinancialImpact module.
 */
export interface FinancialImpact {
  mlr_impact: number;
  mlr_impact_confidence: number;
  stop_loss_exposure: number;
  stop_loss_attachment: number;
  stop_loss_aggregate: number;
  stop_loss_remaining: number;
  reserve_at_risk: number;
  reserve_type: string;
  premium_revenue_risk: number;
  line_of_business: string;
  total_exposure: number;
  total_exposure_confidence: number;
  methodology: string;
  methodology_version: string;
  calculation_timestamp: string;
  sources: any[];
  assumptions: string[];
}

/**
 * Validate RiskObject constraints.
 *
 * @param riskObject - The risk object to validate
 * @throws {Error} If validation fails
 * @returns True if validation passes
 */
export function validateRiskObject(riskObject: RiskObject): boolean {
  // Check score ranges
  if (riskObject.likelihood_score < 0 || riskObject.likelihood_score > 1) {
    throw new Error(`Invalid likelihood_score: ${riskObject.likelihood_score}`);
  }
  if (riskObject.confidence < 0 || riskObject.confidence > 1) {
    throw new Error(`Invalid confidence: ${riskObject.confidence}`);
  }

  // Check required fields
  if (!riskObject.affected_assets || riskObject.affected_assets.length === 0) {
    throw new Error("affected_assets cannot be empty");
  }
  if (!riskObject.business_process_map || riskObject.business_process_map.length === 0) {
    throw new Error("business_process_map cannot be empty");
  }
  if (!riskObject.remediation_owner) {
    throw new Error("remediation_owner cannot be empty");
  }

  return true;
}
